src/Data/Text/Internal.hs

```diff
@@ -124,11 +120,11 @@ firstf _  Nothing      = Nothing
 -- | Checked multiplication.  Calls 'error' if the result would
 -- overflow.
 mul :: Int -> Int -> Int
-#if WORD_SIZE_IN_BITS == 64
-mul a b = fromIntegral $ fromIntegral a `mul64` fromIntegral b
-#else
-mul a b = fromIntegral $ fromIntegral a `mul32` fromIntegral b
-#endif
+mul a b
+  | finiteBitSize (0 :: Word) == 64
+  = fromIntegral $ fromIntegral a `mul64` fromIntegral b
+  | otherwise
+  = fromIntegral $ fromIntegral a `mul32` fromIntegral b
```

Comment:
This introduces some new code. Is there something wrong with keeping the machdeps import and CPP'ing these definitions?

Reply:
GHC constant folds `finiteBitSize (0 :: Word)` branches.

Reply:
There is very little public infrastructure available to test 32 bit platforms, e. g., AFAIK no CI service has such an option. Using `finiteBitSize` allows at least to check that 32 bit code path is well typed, while with CPP it could even do not parse.

Reply:
To my knowledge, a nix one-liner is the easiest way to emulate 32-bit platforms on 64-bit ones. Simply specify the "i686-linux" system in your nixpkgs. Here is a small example illustrating this:

```
$ cat shell.nix
with (import <nixpkgs> { config = {}; overlays = []; system = "i686-linux"; });

mkShell {
  name = "32-bit-nix-emulation-example";

  GHC_INCLUDE_PATH = "${pkgs.haskell.compiler.ghc8104}/lib/ghc-8.10.4/include";

  buildInputs = [
    gcc
  ];
}

$ cat main.c
#include <stdio.h>
#include "MachDeps.h"

int main() {
  printf("%i", WORD_SIZE_IN_BITS);
}

$ nix-shell

nix-shell$ gcc -Wall -I $GHC_INCLUDE_PATH main.c -o main && ./main
32

nix-shell$ exit

$ #edit shell.nix to say `system = x86_64-linux` now

$ gcc -Wall -I $GHC_INCLUDE_PATH main.c -o main && ./main
64
```

TLDR it's sufficient to specify just "system = i686-linux" to emulate a 32-bit platform. I have used this in the past to test things like chronos (the time library) on 32-bit.